node/components/network/src/testonly.rs

```diff
@@ -190,27 +194,66 @@ pub struct InstanceConfig {
 impl Instance {
     /// Constructs a new instance.
     pub fn new(cfg: Config, block_store: Arc<BlockStore>) -> (Self, InstanceRunner) {
-        Self::new_from_config(InstanceConfig {
-            cfg,
-            block_store,
-            attestation: attestation::Controller::new(None).into(),
-        })
+        Self::new_from_config(
+            InstanceConfig {
+                cfg,
+                block_store,
+                attestation: attestation::Controller::new(None).into(),
+            },
+            |_| true,
+            |_, _| SelectionFunctionResult::Keep,
+        )
+    }
+
+    /// Constructs a new instance with custom filter and selection functions for consensus
+    /// component.
+    pub fn new_with_filters(
+        cfg: Config,
+        block_store: Arc<BlockStore>,
+        con_filter_predicate: impl 'static + Sync + Send + Fn(&ConsensusReq) -> bool,
+        con_selection_function: impl 'static
+            + Sync
+            + Send
+            + Fn(&ConsensusReq, &ConsensusReq) -> SelectionFunctionResult,
+    ) -> (Self, InstanceRunner) {
+        Self::new_from_config(
+            InstanceConfig {
+                cfg,
+                block_store,
+                attestation: attestation::Controller::new(None).into(),
+            },
+            con_filter_predicate,
+            con_selection_function,
+        )
     }
 
     /// Construct an instance for a given config.
-    pub fn new_from_config(cfg: InstanceConfig) -> (Self, InstanceRunner) {
-        let (actor_pipe, dispatcher_pipe) = pipe::new();
+    pub fn new_from_config(
+        cfg: InstanceConfig,
+        con_filter_predicate: impl 'static + Sync + Send + Fn(&ConsensusReq) -> bool,
+        con_selection_function: impl 'static
+            + Sync
+            + Send
```

Comment:
oof, that's really ugly. How about we require a preconstructed prunable_mpsc instead?

Reply:
and for the "keep_all" channel (which you use in some tests - please minimize the amount of tests that use it, so that tests are as close to reality as possible), you can add a new constructor to the prunable_mpsc directly, so that the intention is clear (instead of inlining the corresponding predicates in some places).

Reply:
The predicates are defined in the bft crate, so I can't import them into network. What I can do is move the predicates to the concurrency crate (or maybe the roles crate).

Reply:
I can move them into network, but that seems worse...

Reply:
no, I didn't mean the real predicates. Just add
```
pub fn keep_all_channel<T>() -> (Sender<T>, Receiver<T>) {
  channel(||true,|_, _| SelectionFunctionResult::Keep)
}
```
to prunable_mpsc module.

Reply:
That I got, yes. It's already in there. But you asked to minimize the amount of tests that use it. However, all tests in network must use the unpruned channel. I can (and do) use the pruned channel for all tests in bft.

Reply:
ok